### PR TITLE
Feat: roughly achieved reference line & band

### PR DESCRIPTION
**Pull request by scottsut**

### CODE REVIEW

File: webapp/app/components/ListFormLayout/index.tsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * >>
+ */
+
+import ListFormLayout from './ListFormLayout'
+import List from './List'
+import ListItem from './ListItem'
+
+export { ListFormLayout, List, ListItem }
```
Comment (AndyFong) [inline webapp/app/components/ListFormLayout/index.tsx:25]:
```javascript
export { default as ListFormLayout } from './ListFormLayout'
```

File: webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+
+interface IReferenceFormProps {
+  reference: IReference
+  metrics: IDataParamSource[]
+}
+
+interface IReferenceFormStates {
+  prevReference: IReference | null
+  editingReference: IReference | null
+  formValuesToBeSet: Partial<IReference> | null
```
Comment (AndyFong) [inline webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx:75]: remove null check

File: webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+
+  public static getDerivedStateFromProps: GetDerivedStateFromProps<
+    IReferenceFormProps & FormComponentProps,
+    IReferenceFormStates
+  > = (props, state) => {
+    const { reference } = props
+    if (reference !== state.prevReference) {
+      let editingReference = null
+      try {
+        editingReference = JSON.parse(JSON.stringify(reference))
```
Comment (AndyFong) [inline webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx:136]: replace `JSON.parse(JSON.stringify())`

File: webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+    }
+    return null
+  }
+
+  public componentDidMount() {
+    const { reference, form } = this.props
+    if (reference) {
+      const { key, name, ...rest } = reference
+      form.setFieldsValue(rest)
+    }
```
Comment (AndyFong) [inline webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx:154]: can replace with using FormItem initialValue

File: webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+                {getFieldDecorator(
+                  'data.line.color',
+                  {}
+                )(<ColorPicker className={styles.colorPicker} preset />)}
+              </FormItem>
+            </Col>
+          </Row>
+        </>
+      )
+    } else if (type === ReferenceType.Band) {
```
Comment (AndyFong) [inline webapp/app/containers/Widget/components/Workbench/Reference/ReferenceForm.tsx:336]: can split into children components

File: webapp/app/containers/Widget/components/Workbench/Reference/index.tsx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+  public static getDerivedStateFromProps: GetDerivedStateFromProps<
+    IReferenceProps,
+    IReferenceStates
+  > = (props, state) => {
+    const { references } = props
+    if (references && references !== state.prevReferences) {
+      let editingReferences = []
+      let selected = null
+      try {
+        editingReferences = JSON.parse(JSON.stringify(references))
```
Comment (AndyFong) [inline webapp/app/containers/Widget/components/Workbench/Reference/index.tsx:71]: same as above

File: webapp/app/containers/Widget/components/Workbench/Reference/index.tsx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    })
+
+    const modalFooter = [
+      <Button key="cancel" size="large" onClick={onCancel}>
+        取 消
+      </Button>,
+      <Button key="submit" size="large" type="primary" onClick={this.save}>
+        保 存
+      </Button>
+    ]
```
Comment (AndyFong) [inline webapp/app/containers/Widget/components/Workbench/Reference/index.tsx:226]: no need for `modalFooter ` declaration, use Modal default footer instead.

File: webapp/app/containers/Widget/components/Workbench/Reference/types.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ * >>
+ */
+
+import {
+  ReferenceType,
+  ReferenceValueType,
+  ReferenceLabelPosition
+} from './constants'
+
+export type IReference = IReferenceLine | IReferenceBand
```
Comment (AndyFong) [inline webapp/app/containers/Widget/components/Workbench/Reference/types.ts:27]: can use inherit instead

**Comment by scottsut** [comment]
#793

**Pull request merged**